ml/kmeans/KMeans.cc

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "KMeans.h"
+#include <dlib/clustering.h>
+
+void KMeans::train(SamplesBuffer &SB) {
+    std::vector<DSample> Samples = SB.preprocess();
+
+    MinDist = std::numeric_limits<CalculatedNumber>::max();
+    MaxDist = std::numeric_limits<CalculatedNumber>::min();
+
+    {
+        std::lock_guard<std::mutex> Lock(Mutex);
+
+        ClusterCenters.clear();
```

Comment:
wouldn't using past centers from the same dimensions optimize things?
adding data of the same dimension is in most cases not going to move centers much. Therefore I think the k-means algorithm would converge to result faster

Reply:
agree - this is a great optimization idea. Worth doing if easy to implement or one for later if involves a bigger change.

Reply:
> wouldn't using past centers from the same dimensions optimize things?
> adding data of the same dimension is **in most cases** not going to move centers much.

Emphasis mine :)

I could see this delivering some improvements but nowhere near anything that would justify spending the time to verify it's correctness and implement it.

IMHO, the 1st goal after the MVP should be the switch to a (really) online clustering algorithm.

Reply:
Well should have been `majority of cases` but anyway not important now.
In Slack discussion I was going towards direction of online/incremental/streaming clustering algorithm anyway as that would be the best way to do this.
Since this is MVP and you say switching to online clustering algorithm as 1st prio after MVP this optimization would be replaced soon and therefore is not worth it if it takes more than 10 minutes to implement and try.

Reply:
defo worth exploring this as we don't really know on average how many iterations each retraining is "using" - my gut says that it actually would be meaningful. So one i think we can/should revisit later once we can spend time to properly benchmark it and run various experiments. 

main goal here is to get a sort of simple enough first model/implementation thats medium complexity in terms of the ML and build from there on the api and product side - we can then do lots of optimisations on what's going on under the hood to produce those 0/1 anomaly bits so this is all good discussion. 

being able to instrument the performance from a users point of view is key too so that we can at least have a way to try measure future improvements or additions like online approaches (or my favourite which would be to ensemble another model in some way - pca or one class svm or isolation forest) on what matters most which will be user satisfaction with the models. so this is a product discussion then about the "thumbs" data we will want to be able to solicit from users and be able to use that as a stable measure of performance from a user point of view that we can track then as we introduce new options like additional models etc.

Reply:
Resolving this as it's something we can discuss in the future. For the time being, I added a configuration option that allows users to specify the max number of Kmeans iterations.